src/providers/GitHub/index.ts

```diff
@@ -0,0 +1,227 @@
+    const nango = new Nango({ secretKey: process.env.NANGO_SECRET_KEY });
+
+    const integration = (await nango.getIntegration(
+      options.nango_integration_id ?? "github",
+      true, // get credentials
+    )).config as IntegrationWithCreds;
+
+    const connection = await nango.getConnection(
+      options.nango_integration_id ?? "github",
+      options.nango_connection_id,
+    );
+
+    await this.authorize({
+      authStrategy: createOAuthUserAuth,
+      auth: {
+        clientId: integration.client_id,
+        clientSecret: integration.client_secret,
+        clientType: "oauth-app",
+        token: connection.credentials.raw.access_token,
+        scopes: integration.scopes,
+      }
+    })
+  }
+
+  async getDocuments(): Promise<Document[]> {
+    let branchName = this.branch;
+    
+    if (this.branch === undefined) {
+      const repo = await this.octokit.rest.repos.get({
+        owner: this.owner,
+        repo: this.repo,
+      });
+
+      // Not all GitHub repositories have branches.
+      if (repo.data.default_branch === undefined) {
+        throw Error("Could not determine the default branch of the repository. Please specify a branch with the `branch` option.");
+      }
+
+      branchName = repo.data.default_branch;
+    }
+
+    const branch = await this.octokit.rest.repos.getBranch({
+      owner: this.owner,
+      repo: this.repo,
+      branch: branchName,
+    });
+
+    const tree = await this.octokit.rest.git.getTree({
+      owner: this.owner,
+      repo: this.repo,
+      tree_sha: branch.data.commit.sha,
+      recursive: "true",
+    });
+
+    let files = tree.data.tree
+      .filter(item => item.type == "blob");
+
+    if (this.path !== undefined) {
+      files = files.filter(file => {
+        // Check if this.path contains file.path
+        const relative = path.relative(this.path, file.path);
+        return relative && !relative.startsWith("..") && !path.isAbsolute(relative);
+      });
+    }
+
+    if (this.docOnly) {
+      files = files.filter(file => DOC_EXTENSIONS.some(ext => file.path.endsWith(ext)));
+    }
+
+    const blobs = await Promise.all(
+      files.map(async file => {
+        const blob = await this.octokit.rest.git.getBlob({
+          owner: this.owner,
+          repo: this.repo,
+          file_sha: file.sha,
+        });
+
+        return {
+          file,
+          blob: blob.data,
+        }
+      })
+    )
+
+    return blobs.map(({ file, blob }) => ({
+      id: blob.sha,
+      content: blob.content,
+      metadata: {
+        // Construct pretty source URL.
+        sourceURL: `https://github.com/${
+          encodeURIComponent(this.owner)
+        }/${
+          encodeURIComponent(this.repo)
+        }/blob/${
+          encodeURIComponent(branchName)
+        }/${
+          file.path
+            .split("/") // Don't escape slashes, they're a part of the path.
+            .map(part => encodeURIComponent(part))
+            .join("/")
+        }`,
```

Comment:
This is pretty dodgy, but I couldn't find a better way to construct a URL. `file.url` points to an `api.github.com` link, which won't bring up the GitHub UI. This "pretty URL" does.

Reply:
Sounds good, this should be fine.